pyriemann/transfer/pipelines.py

```diff
@@ -0,0 +1,824 @@
+            Set of SPD matrices after rotation step.
+        """
+
+        # used during fit in pipeline, rotate each source domain
+        self.fit(X, y_enc)
+        _, _, domains = decode_domains(X, y_enc)
+        X_rot = np.zeros_like(X)
+        for d in np.unique(domains):
+            idx = domains == d
+            if d != self.target_domain:
+                X_rot[idx] = self.rotations_[d] @ X[idx] @ self.rotations_[d].T
+            else:
+                X_rot[idx] = X[idx]
+        return X_rot
+
+
+class TLEstimator(BaseEstimator):
+    """Transfer learning wrapper for estimators.
+
+    This is a wrapper for any BaseEstimator (i.e. classifier or regressor) that
+    converts extended labels used in Transfer Learning into the usual y array
+    to train a classifier/regressor of choice.
+
+    Parameters
+    ----------
+    target_domain : str
+        Domain to consider as target.
+    estimator : BaseEstimator
+        The estimator to apply on the data points. It can be any regressor or
+        classifier from pyRiemann.
+    domain_weight : None | dict, default=None
+        How to combine the samples from each domain to train the estimator.
+        The dict contains key=domain_name and value=weight_to_assign. If 'None'
+        then assign the same weight to all domains.
+
+    Notes
+    -----
+    .. versionadded:: 0.3.1
+    """
+
+    def __init__(self, target_domain, estimator, domain_weight=None):
+        """Init."""
+        self.target_domain = target_domain
+        self.domain_weight = domain_weight
+
+        if is_regressor(estimator) or is_classifier(estimator):
+            self.estimator = estimator
+        else:
+            raise TypeError(
+                'Estimator has to be either a classifier or a regressor.')
+
+    def fit(self, X, y_enc):
+        """Fit TLEstimator.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_matrices, n_channels, n_channels)
+            Set of SPD matrices.
+        y_enc : ndarray, shape (n_matrices,)
+            Extended labels for each matrix.
+
+        Returns
+        -------
+        self : TLEstimator instance
+            The TLEstimator instance.
+        """
+
+        X_dec, y_dec, domains = decode_domains(X, y_enc)
+
+        if is_regressor(self.estimator):
+            y_dec = y_dec.astype(float)
+
+        if self.domain_weight is not None:
+            w = np.zeros(len(X_dec))
+            for d in np.unique(domains):
+                w[domains == d] = self.domain_weight[d]
+        else:
+            w = None
+
+        if isinstance(self.estimator, Pipeline):
+            sample_weight = {}
+            for step in self.estimator.steps:
+                step_name = step[0]
+                sample_weight[step_name + '__sample_weight'] = w
+            self.estimator.fit(X_dec, y_dec, **sample_weight)
+        else:
+            self.estimator.fit(X_dec, y_dec, sample_weight=w)
+
+        return self
+
+    def predict(self, X):
```

Comment:
When `estimator` is a classifier, the function `predict_proba()` should be available; not when it is a regressor.
How to define this function conditionally? @agramfort @plcrodrigues

Reply:
good catch... we can do black magic but subclassing with a TLClassifier and TLRegressor is the cleanest

Reply:
Ok for subclasses, I will commit this solution.